openid/openid.go

```diff
@@ -0,0 +1,273 @@
+		v.Set("openid.ns.lp", nsTeams)
+		v.Set("openid.lp.query_membership", strings.Join(r.Teams, ","))
+	}
+	if len(r.SRegRequired) > 0 {
+		v.Set("openid.ns.sreg", nsSReg)
+		v.Set("openid.sreg.required", strings.Join(r.SRegRequired, ","))
+	}
+	if len(r.SRegOptional) > 0 {
+		v.Set("openid.ns.sreg", nsSReg)
+		v.Set("openid.sreg.optional", strings.Join(r.SRegOptional, ","))
+	}
+	if (c.Server == usso.UbuntuSSOServer{}) {
+		c.Server = usso.ProductionUbuntuSSOServer
+	}
+	return c.Server.OpenIDURL() + "?" + v.Encode()
+}
+
+// Response contains the values returned from Ubuntu SSO after a
+// successful login.
+type Response struct {
+	// ID contians the claimed_id of the logged in user. This will
+	// always be present in a successful login.
+	ID string
+
+	// Teams contains any launchpad teams that were specified in the
+	// OpenID response.
+	Teams []string
+
+	// SReg contains any simple registration fileds are
+	// were provided in the OpenID response.
+	SReg map[string]string
+}
+
+// verify is used to perform the OpenID verification of the login
+// response. This is declared as a variable so it can be overridden for
+// testing.
+var verify = openid.Verify
+
+// Verify processes a positive assertion from Ubuntu SSO. If the
+// verification is successful any parameters asserted by Ubuntu SSO will
+// be set in the Response. If the OpenID response reports that the login
+// was cancelled then an error will be returned with a cause of
+// ErrCancel. If the OpenID response reports an error occured then an
+// err will be returned with a cause fo ErrOpenID.
+func (c Client) Verify(requestURL string) (*Response, error) {
+	if (c.Server == usso.UbuntuSSOServer{}) {
+		c.Server = usso.ProductionUbuntuSSOServer
+	}
+	if c.NonceStore == nil {
+		c.NonceStore = defaultNonceStore
+	}
+	if c.DiscoveryCache == nil {
+		c.DiscoveryCache = ussoDiscoveryCache{c.Server}
+	}
+
+	u, err := url.ParseRequestURI(requestURL)
+	if err != nil {
+		return nil, err
+	}
+	v := u.Query()
+	switch v.Get("openid.mode") {
+	case "error":
+		return nil, errgo.WithCausef(nil, ErrOpenID, "%s", v.Get("openid.error"))
+	case "cancel":
+		return nil, ErrCancel
+	default:
+		return nil, errgo.Newf("unrecognised mode %q", v.Get("openid.mode"))
+	case "id_res":
+	}
+	if endpoint := v.Get("openid.op_endpoint"); endpoint != c.Server.OpenIDURL() {
+		return nil, errgo.Newf("OpenID response from unexpected endpoint %q", endpoint)
+	}
+
+	// Verify the openid response.
+	id, err := verify(requestURL, c.DiscoveryCache, c.NonceStore)
+	if err != nil {
+		return nil, err
+	}
+	r := Response{
+		ID: id,
+	}
+	// check for extensions in the response.
+	signed := strings.Split(v.Get("openid.signed"), ",")
+	if v.Get("openid.ns.lp") == nsTeams && contains(signed, "lp.is_member") {
+		r.Teams = strings.Split(v.Get("openid.lp.is_member"), ",")
+	}
+	if v.Get("openid.ns.sreg") == nsSReg {
+		r.SReg = make(map[string]string)
+		for k := range v {
+			if !strings.HasPrefix(k, "openid.sreg.") {
```

Comment:
Not really the other signed fields will either be to do with proving the validity of the login, in which case the client doesn't need to know the specifics, or the list of teams which is in another field.